### ISSUE TITLE
Remove createElement

### ISSUE THREAD

**Issue by TylerRHoyer** [Roblox/roact#109]
On #104 I suggested having createElement remove special markers (like Roact.Children, Roact.Ref, events) from the element props and put them into their own keys in the new element. @LPGhatguy said that there is a beauty to having everything inside of the props. So, taking that idea further, I propose putting the component into the props. This has the following effects: 1, There is no need for the elements struct, meaning that only one table instead of two per element need to be created every render. 2, createElement can be removed, preventing a function call for every element recreated every render. 3, The code becomes much more declarative. 4, We lose the ability to track an element's source to a specific line number (We could still track what render method created it though).

When I tested the performance improvement of this creating an element the new way vs calling createElement, I was surprised to find that it was only an 85% improvement. Running
```lua
for i = 1, 1000000 do
local element = createElement("Part")
end
```
took 5934.6 ms and
```lua
for i = 1, 1000000 do
local element = {
    component = "Part"
}
end
```
took 893.5 ms.

When mounting 10000 parts with only their anchored property set, Roact takes 368 ms more than directly creating the parts. We know that the entire process of element creation is overhead and contributes to the 368 ms total overhead of this operation. The total time createElement took was 17 ms. An 85% reduction in createElement time means a 3.9% reduction in the overhead of instance creation using Roact. This % of overhead reduction would be larger for reconciling because it isn't usually as resource intensive but the createElement call remains at the same intensity. These results don't however account for variable number of properties.

Converting code away from using createElement wouldn't be too difficult because most code could be easily converted using a regular expression. Code that uses a variable as the element properties argument would however need to be manually edited. In addition, an edited version of the createElement function could remain in the code base until all code using it has been changed over to using the new method.

**Comment by AmaranthineCodices** [comment]
I'm not sure that the performance gains of removing `createElement` outweigh the constraints it places on us. Right now, the element format is not part of Roact's public API; we can change it however we need to without breaking user code. Switching to direct table creation means:

* We cannot change the format without breaking user code - there is no "version" field that we can use to specify different formats!
* Validation of arguments *must* be deferred to the reconciler
* We can't include debug information automatically, as `createElement` does now:
https://github.com/Roblox/roact/blob/b161fc1556c0bb413f54f2f698b6e00207368dae/lib/createElement.lua#L31

Another thing to consider is that the cost of `createElement` is constant; it doesn't change with the number of properties in the element. The time spent in reconciliation, however, does change - in some cases rather dramatically - with the property count. Eliminating `createElement` calls probably won't have the same performance gains for very large elements, to say nothing of stateful components with lifecycle hooks!

**Comment by TylerRHoyer** [comment]
I don't foresee the additional constraints on the element format being a problem, but we would need to be careful as with any other public API.

Considering that we don't want to pull out different types of properties from the props table, it doesn't make sense to me that we are making elements heavier by pulling out the component type. Making elements a single table rather than two can easily be done without removing `createElement` since as you mentioned, we can easily change the element format.

As for removing/deprecating `createElement`, I'll address the counter arguments in the order they were presented. If we make the decision to make elements a lighter, single table it is very unlikely we will go back or change it to a more complicated structure. Since elements are supposed to be simple, light, and declarative, it is reasonable to assume that createElement will continue to only be used for (a) validation, (b) mapping inputs to properties, and (c) automatically inserting some creation-specific data (like a stack trace).

(a) I brought up this same point in #104 with the descriptions being passed into tasks. @LPGhatguy said that it is more convenient to perform validation at the location the description struct is being used rather than where it is created. I must admit though, it would make sense to using validation at creation if we often created elements and passed them to the reconciler more than once. While it is nitpicky, we could allow the type system to validate the element and give it a seal of validity in the form of a type, which could be easily checked in the reconciler. However, making the creation of every element relatively significantly slower by requiring an additional pre-function call to `createElement` to perform validation to optimize the rare reuse of elements seems to be a poor trade off to me. While it is fine to perform validation in `createElement` while it exists for other reasons, it should not be the sole purpose createElement exists. And as mentioned earlier, a function like `createElement` could be provided to still optimize the rarer case of element reuse.

(b) If the structure of the element is to be a single table, then the only structural changes that could be made would be to the name of values and which values are included. Changing the name of a value could be done through deprecation, but in most cases name changes are minor details and easily avoided with proper planning. We can always add conditional values with the use of symbols and we can always ignore/warn about values we have removed. That being said, the only value elements require right now are the component type, which is always going to be needed (I'd suggest making element.source an optional string, in addition to changing the name to suit, but that is another discussion). We won't ever need to add a required value, unless we changed the very definition of what an element is - in which case we'd have bigger issues.

(c) I'm not entirely sure what other creation specific data we'd want to put into elements beside the stack trace of where it was created, but the good news is that creation specific data will always be optional and can always be added in manually when needed. For the specific case of the stack trace being inserted into elements, other options exists that would allow better debug information globally without increasing overhead by requiring all elements to be created with a special function call. Reporting which component's render was being reconciled in addition to a shallow dump of the element would be sufficient in most cases to find which element was causing the issue. Additional debug information, such as the stack trace of where an element was created, is easy to add and would be outputted with the contents of the element. If a developer really wanted to, they could easily define a function to automate adding creation specific information to an element, which could be found by any optional Roact features that use it. For example, a function which takes a table, inserts a stack trace, and returns the table is four lines of simple code. Removing `createElement` would remove a good place to insert creation specific data into every element at the flip of a flag, but oftentimes specific enough information can be obtained in the reconciler. If not, creation specific data can always be easily inserted. It is not worth slowing down every element creation to allow the flipping of a flag to turn on creation specific information gathering for all elements.

Elements and therefore element creation is meant to be simple and cheap. 85% of the current element creation time could be eliminated with minimal changes to user code. We are currently creating at least two tables for every element when we should be creating only one and performing an unneeded function call. Not to mention that removing the calls to createElement will make code look prettier. ^.^
